### PR TITLE
Optimize custom tpu 2

### PR DESCRIPTION
**Pull request by godmodegalactus**

### CODE REVIEW

File: bench/src/main.rs [hunk 1]
```diff
@@ -39,17 +40,33 @@ async fn main() {
...
 
+    let join_res = join_all(tasks).await;
+
+    let mut run_num = 1;
+    for res in join_res {
+        match res {
+            Ok(metric) => {
+                info!("Run {run_num}: Sent and Confirmed {tx_count} tx(s) in {metric:?} with",);
+                // update avg metric
+                avg_metric += &metric;
```
Comment (aniketfuryrocks) [inline bench/src/main.rs:60]: This is not an accurate way top calculate avg.
cause join_all waits for all futures

File: src/workers/tpu_utils/tpu_connection_manager.rs [hunk 2]
```diff
@@ -116,50 +121,162 @@ impl ActiveConnection {
...
-                already_connected,
-                endpoint.clone(),
-                addr,
-                exit_signal.clone(),
-            )
-            .await else {
-                return None;
-            };
+    // copied from solana code base
+    fn compute_receive_window_ratio_for_staked_node(
```
Comment (aniketfuryrocks) [inline src/workers/tpu_utils/tpu_connection_manager.rs:262]: Is this function not exported from any crate from solana ?

File: src/workers/tpu_utils/tpu_connection_manager.rs [hunk 3]
```diff
@@ -116,50 +121,162 @@ impl ActiveConnection {
...
-                    connection.open_uni(),
-                )
-                .await
-                {
-                    Ok(Ok(unistream)) => return Some(unistream),
-                    Ok(Err(_)) => (),
-                    Err(_) => return None,
+        last_stable_id: Arc<AtomicU64>,
+    ) {
+        for _ in 0..3 {
```
Comment (aniketfuryrocks) [inline src/workers/tpu_utils/tpu_connection_manager.rs:207]: What's the 3 for. Can be made a constant

**Review by aniketfuryrocks — CHANGES_REQUESTED** [review]

File: bench/src/main.rs (hunk 1 above)
Comment (godmodegalactus) [reply bench/src/main.rs:60]: Ok i will update it later.

File: src/workers/tpu_utils/tpu_connection_manager.rs (hunk 2 above)
Comment (godmodegalactus) [reply src/workers/tpu_utils/tpu_connection_manager.rs:262]: It is not public but private function.

**Pull request merged**